kimchi/src/index.rs

```diff
@@ -148,28 +156,34 @@ pub fn constraints_expr<F: FftField + SquareRootField>(
     domain: D<F>,
     chacha: bool,
     dummy_lookup_value: Option<&[F]>,
+    powers_of_alpha: &mut alphas::Builder,
 ) -> E<F> {
-    let expr = poseidon::constraint();
-    let expr = expr + varbasemul::constraint(super::range::MUL.start);
-    let (alphas_used, complete_add) = complete_add::constraint(super::range::COMPLETE_ADD.start);
-    assert_eq!(alphas_used, super::range::COMPLETE_ADD.len());
-    let expr = expr + complete_add;
-    let expr = expr + endosclmul::constraint(2 + super::range::ENDML.start);
-    let expr = expr + endomul_scalar::constraint(super::range::ENDOMUL_SCALAR.start);
-
-    let expr = if let Some(dummy) = dummy_lookup_value {
-        let constraints = lookup::constraints(dummy, domain);
-        let combined = Expr::combine_constraints(2 + super::range::CHACHA.end, constraints);
-        expr + combined
-    } else {
-        expr
-    };
+    // gates
+    let alphas = powers_of_alpha.register(ConstraintType::Gate, 21);
+
+    let mut expr = poseidon::constraint(alphas.clone().take(15));
```

Comment:
I was pretty limited previously because alpha was defined in the protocol crate, and since the circuit crate was a dependency it couldn't depend on the alpha type directly. Now that I moved the circuit crate into the kimchi crate, this should not be a limitation anymore. You can chime in https://github.com/o1-labs/proof-systems/issues/272, but I think this could be a good change in the future

Reply:
Oooh that makes sense! Dependencies huh.... Fine, I'll check out that issue :)